### ISSUE TITLE
Can't build it successfully

### ISSUE THREAD

**Issue by S0ar1** [darchr/AutoTM#2]
Hi sir, Now I tried to follow your work with Optane and dl. But I can't build the AutoTM successfully now. " Failed to precompile nGraph" is shown even I build it follow your guide without mistake. 
I wonder if it is because the ngraph has been moved into openvino? Or it is just my fault.
Plz let me know, Thank you!

ERROR: LoadError: Failed to precompile nGraph [96f067b3-e350-5ae7-90f2-8283c539009d] to /root/.julia/compiled/v1.2/nGraph/XNBRd.ji.

**Comment by hildebrandmw** [comment]
Hi @S0ar1 , thank you for your interest! Unfortunately, the project has undergone some bit-rot over the last few years. A while ago, I tried to get this working again and ran into the following issues:

* We're using a fork of nGraph, so aren't necessarily affected by the move to OpenVINO. Nevertheless, standard library header changes caused our fork to no longer build properly (this is remedied by including the accidentally omitted headers, but this only gets us so far).
* If I recall correctly, the code was written for an older version of Julia. At that point, the handling for external dependencies wasn't as robust as it is now. As such, the version of CxxWrap that we depend on basically required building the necessary version of Julia from source.
* While building that version of Julia, I encountered kind of a hard-stop when building OpenBLAS (a dependency of Julia). I'm still not entirely sure what was happening there, but it appeared that the architecture detection wasn't working correctly for the hardware we were running on. (I am far from an expert build engineer, someone with more experience would probably get much further).
* I never made it far enough to start getting Gurobi working again.

Basically, there are too many moving parts and too little incentive to actively maintain the code to keep the project in a working state. :disappointed:

I will pin this issue and leave a note on the README communicating the state of the project.

I sincerely apologize that this is situation and hope you did not lose too much time dealing with this. If you have any questions regarding the approach or related to the project in general, please feel free to ask and I will do what I can to assist.

Regards,
Mark

**Comment by S0ar1** [comment]
> 
HI Mark! Thank you for your earnest reply!
About the 1st point. Could you explain "this is remedied by including the accidentally omitted headers" in more details? I am confused with it. 
 In addition,I notice that you recommend 10.1 cuda. But I use the 11.6 version.
 Is it possible that the distinguish of cuda cause the fault of the build ?

Regards,
Soar

**Comment by hildebrandmw** [comment]
Hi Soar,

Apologies for the delay! About including the headers: we used a custom fork of `ngraph` for this project that added the Optane support and kernels supporting move operations (both DRAM <-> Optane and CPU DRAM <-> GPU DRAM). At some point as the C++ standard libraries used by compilers were updated, this fork stopped building. The reason for this is that as the standard library changes, the headers transitively included may change (for example, using `#include <sstream>` might also pull in `<string>` in one version, but not in another). The headers for the `ngraph` library were accidentally depending on some transitive includes which broke. I forget if I fixed this or not (the fix was as simple as including the headers that were missed).

For CUDA, I would be mildly surprised if CUDA 11.6 simply worked out of the box. Unfortunately, that is far from the only issue in getting this project to run again 😞.
